### 零阶差商

$$f[x_0] = y_0$$

### 一阶差商

$$f[x_i, x_j] = \frac{y_j - y_i}{x_j - x_i}$$

### 二阶差商

$$f[x_i, x_j, x_k] = \frac{\frac{y_k- y_j]}{x_k - x_j} - \frac{y_j - y_i}{x_j - x_i}}{x_k - x_i} = \frac{f[x_j, x_k] - f[x_i, x_j]}{x_k-x_i}$$

### n阶差商

$$f[x_0, x_1, ..., x_n] = \frac{f[x_1, ..., x_{n}] - f[x_0, ..., x_{n-1}]}{x_n - x_0}$$

In [12]:
import pandas as pd
from sympy import *

x = [0, 30, 45, 60, 90]
y = [0, 0.5, 0.70711, 0.86603, 1]

df = pd.DataFrame({"x值": x, "0阶差商":y})

for j in range(df.shape[0]-1):
    for i in range(df.shape[0]-1):
        df.loc[i+1, str(j+1)+"阶差商"] = (df.iloc[i+1, j+1] - df.iloc[i, j+1])/(df.iloc[i+1, 0] - df.iloc[i-j, 0])

df.fillna(' ')

x值     0阶差商        1阶差商         2阶差商         3阶差商         4阶差商
0   0  0.00000                                                   
1  30  0.50000   0.0166667                                       
2  45  0.70711   0.0138073 -6.35407e-05                          
3  60  0.86603   0.0105947 -0.000107089 -7.25802e-07             
4  90  1.00000  0.00446567   -0.0001362 -4.85185e-07  2.67353e-09

由牛顿插值公式可得插值多项式
$$N_1(x) = 0 + 0.0166667(x - 0)$$

$$N_2(x) = N_1(x) + -6.35407 \times 10^{-5}(x - 0)(x - 30)$$

$$N_3(x) = N_2(x) + -7.25802 \times 10^{-7}(x - 0)(x - 30)(x - 45)$$

$$N_4(x) = N_3(x) + -2.67353 \times 10^{-9}(x - 0)(x - 30)(x - 45)(x - 60)$$

In [14]:
x = 40
N = 0
for i in range(4):
    w = 1
    for j in range(i+1):
        w *= (x - df.iloc[j, 0])
    N += df.iloc[i+1, i+2] * w
    print('N' + str(i+1) + '(40) = ' + f'{N}')

N1(40) = 0.6666666666666666
N2(40) = 0.6412503703703704
N3(40) = 0.6427019753086419
N4(40) = 0.642808916323731


故$sin40^o \approx 0.6428089$

### 误差

$$R_n(x) = \frac{f^{n+1}(\xi)}{(n+1)!}\omega_{n + 1} = \frac{f^{n+1}(\xi)}{(n+1)!} \prod_{i=0}^n (x - x_i)$$


$$R_5(x) = \frac{cos(x)}{5!}(x - 0)(x - 30)(x - 45)(x - 60)(x - 90) \leq \frac{1}{5!}(x - 0)(x - 30)(x - 45)(x - 60)(x - 90) * (\frac{\pi}{180})^5$$

后面的$(\frac{\pi}{180})^5$是度转弧度


In [24]:
from sympy import *

x = Symbol('x')
f = sin(x)
print(f"sin(x)的5阶导数为： {diff(f, x, 5)}")


# 阶乘
jc  = 5 * 4 * 3 * 2 * 1
# 误差
R =  1 / jc * (x - 0) *(x - 30) * (x - 45) * (x - 60) * (x - 90) * (pi/180) ** 5
print(f"误差为: {R.evalf(subs={x: 40})}")


sin(x)的5阶导数为： cos(x)
误差为: -0.0000269920324632651
